Make deployment on cloud.gov easy

mogul opened the pull request:
In order to justify more spike stories (and win over skeptics on diagram-as-config among government Python coders), a portfolio architect (me) demonstrate a fast, compliant path to production for spiff-arena by deploying and demoing it on the [cloud.gov](https://cloud.gov) PaaS.

## Acceptance Criteria

- GIVEN  I have logged into cloud.gov from the CLI
    AND  I have cloned this repository
    WHEN I copy `vars.yml-template` to `vars.yml`
    AND  I replace `slug` with a custom value
    AND  I run `cf push --vars-file vars.yml`
    AND  I open the URL `https://spiffworkflow[[slug]].app.cloud.gov`
    THEN I see the SpiffWorkflow login screen

---

For demo purposes:
* [x] Show it populated with the examples from the upstream repository
* [x] Show creating a ServiceAccount associated with a user, getting the api_key, and then reproducing [the Web API example](https://www.youtube.com/watch?v=6fymiAZuRk4) live
* [x] Show how to interact with dmn tables

For production-readiness:
* [x] Show it using Postgres database instead of sqlite
* [ ] Show it using the cloud.gov IdP service

mogul committed:
First whack at trying to deploy with the Python buildpack

Currently erroring out with:

```
   2024-02-29T13:05:03.89-0800 [APP/PROC/WEB/0] ERR Traceback (most recent call last):
   2024-02-29T13:05:03.89-0800 [APP/PROC/WEB/0] ERR File "/home/vcap/app/bin/save_all_bpmn.py", line 4, in <module>
   2024-02-29T13:05:03.89-0800 [APP/PROC/WEB/0] ERR from spiffworkflow_backend import create_app
   2024-02-29T13:05:03.89-0800 [APP/PROC/WEB/0] ERR ModuleNotFoundError: No module named 'spiffworkflow_backend'
   2024-02-29T13:05:03.89-0800 [APP/PROC/WEB/0] ERR Exited with BAD EXIT CODE '1' in bin/boot_server_in_docker script at line: 63.
   2024-02-29T13:05:03.90-0800 [APP/PROC/WEB/0] OUT Exit status 1
```

asteel-gsa committed:
Attempt to build image for later use

asteel-gsa committed:
Attempt to build image for later use

asteel-gsa committed:
attempt to build again

asteel-gsa committed:
re-enable slack notify

won't actually do anything since we dont have a webhook enabled

asteel-gsa committed:
Use our images

asteel-gsa committed:
move jq into `base as deployment`

asteel-gsa committed:
move jq into `base as deployment`

mogul committed:
Switch from buildpack to image, set required variables

We're not tackling how to parse postgres service details out of
VCAP_SERVICES just yet, sticking with sqlite for now.

mogul commented:
Where we're stuck: As of now we _think_ we have set all the requisite [variables for the frontend](https://github.com/GSA-TTS/spiff-arena/blob/41cce00f22cae72fdf054434ce10067e3f411776/manifest.yml#L19-L23) and [variables for the backend](https://github.com/GSA-TTS/spiff-arena/blob/41cce00f22cae72fdf054434ce10067e3f411776/manifest.yml#L65-L97) correctly, and [the backend status is reporting `200` (live link)](https://spiffworkflow-groundhog-heartbeat.app.cloud.gov/api/v1.0/status). However, when we try to [visit the frontend (live link)](https://spiffworkflow-groundhog-heartbeat.app.cloud.gov), we are redirected to `https://spiffworkflow-groundhog-heartbeat.app.cloud.gov/api/v1.0/login?redirect_url=https://spiffworkflow-groundhog-heartbeat.app.cloud.gov/&authentication_identifier=default` and we see 

```
{
  "error_code": "internal_server_error",
  "message": "InvalidRedirectUrlError: Invalid redirect url was given: 'https://spiffworkflow-groundhog-heartbeat.app.cloud.gov/'. It must match the domain the frontend is running on.",
  "status_code": 500
}
```

🤔🤔🤔

mogul committed:
Make parallel deployments easier

* Rename stuff to be clearer that the "slug" isn't random
* Include the slug in the app name to facilitate parallel deployments
    in the same space.
* Default to 1 image, so you can fit two deployments in a 1G sandbox
    space

mogul commented:
Just redeployed after rebasing to main, and we got a better error message:

```
{
  "error_code": "internal_server_error",
  "message": "InvalidRedirectUrlError: Invalid redirect url was given: 'https://spiffworkflow-groundhog-heartbeat.app.cloud.gov/'. It must start with the frontend url: 'https://spiffworkflow-groundhog-heartbeat.app.cloud.gov/api'",
  "status_code": 500
}
```

mogul commented:
Solved in the Spiff Discord!
![image](https://github.com/GSA-TTS/spiff-arena/assets/47762/1d206e74-8eda-431d-ba5b-e1a76663d182)

mogul committed:
Correctly parse variable name

This is asking for the URL to the frontend that the backend should use,
not vice-versa!

mogul commented:
(I'm tearing that instance down now since the route was published and I don't want people monkeying with it while we work.)

mogul commented:
Current status: I am redirected to the login page, but when I try to log in with the default creds I'm told I have an invalid token... but! this is progress!
```
{
  "error_code": "invalid_token",
  "message": "Login failed. Please try again",
  "status_code": 401
}
```

mogul commented:
Current status: We were having trouble successfully logging in. I bought that back to the SpiffWorkflow Discord and they reproduced the problems we were having locally on a path-based setup. They [made a fix](https://github.com/sartography/spiff-arena/pull/1332) and [better documented how to host the API on a subdomain](https://github.com/sartography/spiff-arena/blob/main/docs/DevOps_installation_integration/path_based_routing.md) upstream. I think Cloud Foundry does the proxying/path removal as they expect, so we should be good there. We need to rebase our tree and generate a new image to see if their changes resolved the problem we're seeing!

asteel-gsa commented:
`main` sync'd w/ upstream, branch rebased onto `main`. 
Image [Building Now](https://github.com/GSA-TTS/spiff-arena/actions/runs/8560837977)

mogul commented:
>  I think Cloud Foundry does the proxying/path removal as they expect, so we should be good there.

Looks like I was wrong; [CF doesn't strip the path element from requests delivered to an app that's on a subpath of a route](https://docs.cloudfoundry.org/devguide/deploy-apps/routes-domains.html#:~:text=Cloud%20Foundry%20does%20not%20route%20requests%20for%20context%20paths%20to%20the%20root%20context%20of%20an%20app.%20Apps%20must%20serve%20requests%20on%20the%20context%20path.).

mogul commented:
We should be able to strip the path from the request via judicious application of nginx to implement a proxy. Three options for that spring to mind:

1. If we ensure nginx is installed in the spiffworkflow image, I _think_ we can do that by moving the invocation of the `spiffworkflow-background` app's `start-command` to [a sidecar command in the manifest](https://docs.cloudfoundry.org/devguide/sidecars.html#create), and then making nginx the main command. However, I've never tried making a sidecar with a Docker image app...
2. If we move from deploying spiffworkflow as a Docker image app to deploying it as a `python_buildpack` app, then we can definitely go the sidecar route... but at that point we're back in the territory of trying to work out things that are already solved in the upstream docker-compose.yml path, which was why we stopped work on the buildpack path earlier in this branch. (We may still need to revisit that path for production in any case but I don't want to tackle that until we successfully run spiffworkflow in our environment, so we can unblock the FAC team's adoption of it.)
3. We can deploy `spiffworkflow-backend` on an internal route, and then deploy a separate proxy app on the `/api` subpath of the public route using the `nginx_buildpack`.

Having written these out, I think that the third option is the most straightforward in terms of solving the specific problem we have right now without introducing unknowns into the `spiffworkflow` side, where we're still getting our bearings.

mogul commented:
There's an [upstream PR](https://github.com/sartography/spiff-arena/pull/1340) just now:
>  if backend is generating urls, we need to be careful. perhaps we try to generate a URL that mimics the current url (/api prefix if we are being accessed at /api and / if not)

We should see what happens with it!

mogul commented:
> There's an [upstream PR](https://github.com/sartography/spiff-arena/pull/1340) just now:
> [...]
> We should see what happens with it!

[Huzzah!](https://github.com/sartography/spiff-arena/pull/1344)

mogul committed:
Use new method for subpath prefixing

See https://github.com/sartography/spiff-arena/pull/1344

mogul committed:
Turn down the log noise a bit

mogul commented:
The upstream change was still generating bad URLs. I've got [a PR in flight](https://github.com/sartography/spiff-arena/pull/1349) that should fix this sub-path case. I merged that branch into this branch so I can test it with our images, [once they're built](https://github.com/GSA-TTS/spiff-arena/actions/runs/8591555529).

mogul commented:
Now chasing sartography/spiff-arena#1350. [Image build in progress.](https://github.com/GSA-TTS/spiff-arena/actions/runs/8593821952?pr=1)

mogul committed:
Preserve https when generating redirects

mogul commented:
Having merged in the pending branch from upstream, current status:
![image](https://github.com/GSA-TTS/spiff-arena/assets/47762/4ac72674-d6c8-429a-a1df-e8b879f76143)
😎 😎 😎

mogul commented:
Thinking now about what I would want to add to this before we present on it...

For demo purposes:
* [x] Show it populated with the examples from the upstream repository
* [ ] Show creating a ServiceAccount associated with a user, getting the api_key, and then reproducing [the Web API example](https://www.youtube.com/watch?v=6fymiAZuRk4) live

For production-readiness:
* [x] Show it using Postgres database instead of sqlite
* [ ] Show it using the cloud.gov IdP service

mogul commented:
@asteel-gsa just rebased and force-pushed... Can you confirm this is working for you too?

mogul commented:
I should be clearer... It works for me to login, but I'm still getting tons of 502s and non-functional pages. Are you also able to at least login?

mogul committed:
Increase memory for each app to 512M

Enough to function properly, but still fits in a sandbox account

mogul commented:
> I'm still getting tons of 502s and non-functional pages.

This was an OOM situation... The frontend and backend need more than the 256M each we'd given them, and while cloud.gov restarted either of them we'd see blank pages and errors in the Chrome Inspector console. Now they're at 512M each (still small enough to deploy in a cloud.gov sandbox account) and everything seems to be working.

asteel-gsa commented:
> @asteel-gsa just rebased and force-pushed... Can you confirm this is working for you too?

I am only able to access `https://spiffworkflow((slug)).app.cloud.gov/api/v1.0/ui/` and am unable to get to the login screen for the frontend with latest. Probably best to touch base and figure out where the diff is, so we can document.

asteel-gsa commented:
> > @asteel-gsa just rebased and force-pushed... Can you confirm this is working for you too?
> 
> I am only able to access `https://spiffworkflow((slug)).app.cloud.gov/api/v1.0/ui/` and am unable to get to the login screen for the frontend with latest. Probably best to touch base and figure out where the diff is, so we can document.

Figured out what needed to be done
`cf bind-security-group public_networks_egress ORGNAME --lifecycle running --space SPACENAME` which gave the `public_networks_egress` to my sandbox space, allowing the redirects to occur when going to the frontend, allowing me to see the login splash and use the default credentials to gain access to the application.

asteel-gsa committed:
Update jq to be in the apt-get sequence

asteel-gsa committed:
TODO: Document

mogul committed:
Only the backend needs a database binding

mogul committed:
Persist across restarts

Use the bound Postgres service instead of SQLite

asteel-gsa committed:
set the defaultBranch to main for testing bpmn examples

asteel-gsa commented:
Figured out how to get the process-models "examples" to show up..

inside `/bin/clone_process_models.sh` there is a command 
```
git clone -b "$SPIFFWORKFLOW_BACKEND_GIT_SOURCE_BRANCH" "$SPIFFWORKFLOW_BACKEND_GIT_PUBLISH_CLONE_URL" "$SPIFFWORKFLOW_BACKEND_BPMN_SPEC_ABSOLUTE_DIR"
```

Which, seems to have worked to drop the git fork we wanted into `/app/process_models/`, which gave us this.
![image](https://github.com/GSA-TTS/spiff-arena/assets/130377221/056f2722-7091-41fe-94ca-609053785eda)
![image](https://github.com/GSA-TTS/spiff-arena/assets/130377221/d30b2f45-54a8-4375-9c86-7a5be0b99b06)

I did however need to `rm -rf /app/process_models` as this seems to be part of the `/app/bin/boot_server_in_docker` to `git init` [here](https://github.com/GSA-TTS/spiff-arena/blob/72f039fde0f743ee49fa310b50d326e869b0d13c/spiffworkflow-backend/bin/boot_server_in_docker#L83-L87)


Not entirely sure why directly calling this script wont work? It seems executable in the repo, but has no output at all.

Attempted to add it to the command block with no execution. Going to mess around with it, but this is the current blocker for me.
```
command: |
/app/bin/boot_server_in_docker
/app/bin/clone_process_models
```

asteel-gsa committed:
Disable git init on boot_server_in_docker to test process_models examples loading

asteel-gsa commented:
> Attempted to add it to the command block with no execution. Going to mess around with it, but this is the current blocker for me.

Figured it out. (i think). Current stash has the working version, but hasn't been commited, since, for whatever reason @mogul with the latest commits you did, i get a crash loop on attempting to do the connection to the database, though the `uri` does look correct. Something with `sqlalchemy` is still a bit weird for the `postgres` instance. If you send me your `aws_rds` `cf service` plan, and it works for you, I can tear down my `db_instance` and retry, but for now, my stash is using `sqllite` db.

Stash References:
`/bin/boot_server_in_docker`

```bash
# ignore this stash, i was testing something, but of course, as I pushed and was about to wait for a build on the backend, this was unnessary
- # git init "${SPIFFWORKFLOW_BACKEND_BPMN_SPEC_ABSOLUTE_DIR}"
+ git init "${SPIFFWORKFLOW_BACKEND_BPMN_SPEC_ABSOLUTE_DIR}"
git config --global --add safe.directory "${SPIFFWORKFLOW_BACKEND_BPMN_SPEC_ABSOLUTE_DIR}"
```

`spiffworkflow_backend/config/local_development.py`
```python
- SPIFFWORKFLOW_BACKEND_GIT_USERNAME = "sartography-automated-committer"
- SPIFFWORKFLOW_BACKEND_GIT_USER_EMAIL = f"{SPIFFWORKFLOW_BACKEND_GIT_USERNAME}@users.noreply.github.com"
+ SPIFFWORKFLOW_BACKEND_GIT_USERNAME = environ.get(
    "SPIFFWORKFLOW_BACKEND_GIT_USERNAME",
    default="sartography-automated-committer")
+ SPIFFWORKFLOW_BACKEND_GIT_USER_EMAIL = environ.get(
    "SPIFFWORKFLOW_BACKEND_GIT_USER_EMAIL",
    default=f"{SPIFFWORKFLOW_BACKEND_GIT_USERNAME}@users.noreply.github.com"
)
```

`manifest.yml`

```yml
  command: |
    # Get the postgres URI from the service binding. (SQL Alchemy insists on "postgresql://".🙄)
    export SPIFFWORKFLOW_BACKEND_DATABASE_URI=$( echo $VCAP_SERVICES | jq -r '.["aws-rds"][].credentials.uri' | sed -e s/postgres/postgresql/ ) 
    # export SPIFFWORKFLOW_BACKEND_DATABASE_URI=$( echo $VCAP_SERVICES | jq -r '.["aws-rds"][].credentials.uri' | sed -e s/postgres/postgresql/ )
    /app/bin/clone_process_models
    /app/bin/boot_server_in_docker

# VCAP_SERVICES
SPIFFWORKFLOW_BACKEND_DATABASE_URI: "sqlite:///db.sqlite3"

# Rearrange later
# https://github.com/sartography/spiff-arena/blob/293aa867a1cef056c5bee3ef037be31047fdc49e/spiffworkflow-backend/src/spiffworkflow_backend/config/default.py#L157-L179
SPIFFWORKFLOW_BACKEND_GIT_PUBLISH_CLONE_URL: ((git-process-models))
SPIFFWORKFLOW_BACKEND_GIT_USERNAME: "asteel-gsa"
SPIFFWORKFLOW_BACKEND_GIT_USER_EMAIL: "130377221+asteel-gsa@users.noreply.github.com"
SPIFFWORKFLOW_BACKEND_GIT_COMMIT_ON_SAVE: "true"
SPIFFWORKFLOW_BACKEND_GIT_SOURCE_BRANCH: "main"
SPIFFWORKFLOW_BACKEND_BPMN_SPEC_ABSOLUTE_DIR: "/app/process_models"
SPIFFWORKFLOW_BACKEND_GIT_SSH_PRIVATE_KEY: "/tmp/ssh_private_key.XXXXXX"

# https://github.com/sartography/spiff-arena/blob/main/spiffworkflow-backend/bin/clone_process_models
# https://github.com/sartography/spiff-arena/blob/main/spiffworkflow-backend/bin/find_sample_process_models


SPIFFWORKFLOW_BACKEND_DATABASE_TYPE: "sqlite"
# SPIFFWORKFLOW_BACKEND_DATABASE_TYPE: "postgres"
```

`vars.yml / vars.yml-template`
```yaml
git-process-models: https://github.com/gsa-tts/gsa-process-models.git
```

Live Link: [Here](https://spiffworkflow-sun-moon.app.cloud.gov/)

Ill keep the stash until we figure out the postgres issue, then commit it up, but adding the code for reference for what needed to be done @mogul

asteel-gsa commented:
![image](https://github.com/GSA-TTS/spiff-arena/assets/130377221/69926d4d-48d8-4c1c-9496-739986339186)
```
['/app/src/spiffworkflow_backend/../../bin/git_commit_bpmn_models_repo', '/app/process_models', 'User: admin@spiffworkflow.org added process group testing-process-group', 'main']Stdout: [main e551b47] User: admin@spiffworkflow.org added process group testing-process-group
 1 file changed, 5 insertions(+)
 create mode 100644 testing-process-group/process_group.json
Stderr: fatal: could not read Username for 'https://github.com': No such device or address
Exited with BAD EXIT CODE '128' in /app/src/spiffworkflow_backend/../../bin/git_commit_bpmn_models_repo script at line: 46.
```

Doesn't like my user/email, redeploying with the default to test 
EDIT: Didn't like it, still messing around with it.

asteel-gsa committed:
add myself to example.yml to see if process creation can commit

asteel-gsa commented:
Current status of process models:

- Holding them in local stash for now, but they are working.
- Determined that `SPIFFWORKFLOW_BACKEND_GIT_PUBLISH_TARGET_BRANCH` is not something we need immediately, but this is what I expected to be the target branch for the push when creating a new process group
- `SPIFFWORKFLOW_BACKEND_GIT_SOURCE_BRANCH` was originally set to `main`, however, i made a second branch `publish` that is now the placeholder for saved models, awaiting PR.
- `SPIFFWORKFLOW_BACKEND_GIT_PUBLISH_CLONE_URL` needs to be `ssh compatible`, i.e in `vars.yml`
```yaml
git-process-models: git@github.com:GSA-TTS/gsa-process-models.git
```
- This also needs a git ssh key generated with `ssh-keygen -t rsa -b 4096 -C "my-git-email@my-git-email.blah"` stored as `ssh-key` in `manifest.yml`
```yaml
ssh-key: |
  -----BEGIN OPENSSH PRIVATE KEY-----
  . . . 
  -----END OPENSSH PRIVATE KEY-----
```
- Created my own user in `examples.yml` see https://github.com/GSA-TTS/spiff-arena/pull/1/commits/ea5c6ccbef04daacac4c60c0ca4598a7dd7c3977 for my own testing but beared no weight on the actual commit process

![image](https://github.com/GSA-TTS/spiff-arena/assets/130377221/15b1c643-2735-43e4-bf8a-b3b323e5b96f)

### Final Thoughts
- We will need a "Service Account" in the GSA-TTS space that is used to handle these commits, so they are not tethered to a single specific users SSH key. I generated a key for testing purposes and it is considered secret with no intention to share. When deploying this to a new users space, they will need to generate a key for themselves, and add it to git.

mogul commented:
> * Determined that `SPIFFWORKFLOW_BACKEND_GIT_PUBLISH_TARGET_BRANCH` is not something we need immediately, but this is what I expected to be the target branch for the push when creating a new process group
> * `SPIFFWORKFLOW_BACKEND_GIT_SOURCE_BRANCH` was originally set to `main`, however, i made a second branch `publish` that is now the placeholder for saved models, awaiting PR.

I think this kind of makes sense... The flow is  from `source -> publish` branches. Clearly the naming could be better (eg "target" instead of "publish").

> * We will need a "Service Account" in the GSA-TTS space that is used to handle these commits, so they are not tethered to a single specific users SSH key. I generated a key for testing purposes and it is considered secret with no intention to share. When deploying this to a new users space, they will need to generate a key for themselves, and add it to git.

I think it's right that commits will be attributable to specific users. I think the idea is that you run SpiffWorkflow locally (via docker-compose) or in your sandbox space, configured with your own personal SSH key and the branch name that you want to work with, and edits you make there are pushed to that branch. Then when you're ready that branch gets reviewed in a PR, and merged to `main`. Then that change is promoted across `dev`, `preview`, `staging`, `prod`, etc just as we do with any other code change.

mogul commented:
> I think the idea is that you run SpiffWorkflow locally (via docker-compose)

This especially makes sense when you're developing changes and adding tests for a site that uses SpiffWorkflow via API at the same time.

asteel-gsa commented:
> > * Determined that `SPIFFWORKFLOW_BACKEND_GIT_PUBLISH_TARGET_BRANCH` is not something we need immediately, but this is what I expected to be the target branch for the push when creating a new process group
> > * `SPIFFWORKFLOW_BACKEND_GIT_SOURCE_BRANCH` was originally set to `main`, however, i made a second branch `publish` that is now the placeholder for saved models, awaiting PR.
> 
> I think this kind of makes sense... The flow is from `source -> publish` branches. Clearly the naming could be better (eg "target" instead of "publish").
> 
> > * We will need a "Service Account" in the GSA-TTS space that is used to handle these commits, so they are not tethered to a single specific users SSH key. I generated a key for testing purposes and it is considered secret with no intention to share. When deploying this to a new users space, they will need to generate a key for themselves, and add it to git.
> 
> I think it's right that commits will be attributable to specific users. I think the idea is that you run SpiffWorkflow locally (via docker-compose) or in your sandbox space, configured with your own personal SSH key and the branch name that you want to work with, and edits you make there are pushed to that branch. Then when you're ready that branch gets reviewed in a PR, and merged to `main`. Then that change is promoted across `dev`, `preview`, `staging`, `prod`, etc just as we do with any other code change.

Ah! That makes a lot of sense then.

asteel-gsa commented:
> `cf bind-security-group public_networks_egress ORGNAME --lifecycle running --space SPACENAME`

also needed to do this for postgres in sandbox

```
cf bind-security-group trusted_local_networks ORGNAME  --lifecycle running --space SPACENAME
cf bind-security-group trusted_local_networks_egress ORGNAME  --lifecycle running --space SPACENAME
```

asteel-gsa committed:
The process model commit

- Process model examples are shown
- Process model edits are commited to repo
- Process model publishing works

asteel-gsa committed:
Update Cloudgov Readme

mogul commented on README-cloudgov.md:
```diff
@@ -1,11 +1,20 @@
+### Generating a Github SSH Key
+In order to utilize some of the processes in this application, we utilize a forked process model repo, accessible via ssh endpoint.
+`git@github.com:GSA-TTS/gsa-process-models.git`.
+Generate a new ssh key pairing for use with git, using `ssh-keygen -t rsa -b 4096 -C "my-git@email.blah"`. Once this has been created, navigate to the location where the keys are stored, usually `~/.ssh/` and copy the `<my_key_name>.pub` to [Github SSH Keys](https://github.com/settings/keys). Then, copy the plaintext `<my_key_name>` private key to `vars.yml` under the variable `github_ssh_key`. A ...
```
> Alternatively, if you wish to use the actual keyfile and copy it onto the deployed instance for use, you can use the environment var `SPIFFWORKFLOW_BACKEND_GIT_SSH_PRIVATE_KEY_PATH:`

I don't think this will work... The path would have to be target-side, and we don't have any process that would cause that local file to land in the container created from the GCR image, right?

asteel-gsa replied:
> > Alternatively, if you wish to use the actual keyfile and copy it onto the deployed instance for use, you can use the environment var `SPIFFWORKFLOW_BACKEND_GIT_SSH_PRIVATE_KEY_PATH:`
> 
> I don't think this will work... The path would have to be target-side, and we don't have any process that would cause that local file to land in the container created from the GCR image, right?

This was more of a reference, if we want to handle this via Github Action later down the road. We could "theoretically" deploy with that env var, store a key in Github Secrets and do some bash to write it to the key locally. But I can remove it if we want.

mogul committed:
Make the template match the example

mogul commented:
Hmm, I'm actually not able to see a process run all the way through. 

I started an instance of `Examples > Minimal Example`. It took me to the "Show Example Manual Task" step, but when I clicked `Continue` I just saw the pink `Error:` banner appear at the top of the screen:
![image](https://github.com/GSA-TTS/spiff-arena/assets/47762/eba768e8-464a-46d4-a147-974d282e5b71)

If I click `Continue` again, the error message says that the process instance has already completed:
![image](https://github.com/GSA-TTS/spiff-arena/assets/47762/ac4e7f58-eaa8-4630-9a11-52d9ca4863e2)

And that is what I see when I click to the details...
![image](https://github.com/GSA-TTS/spiff-arena/assets/47762/4b378016-8021-4a2f-b838-feba5dd45c99)

I didn't see any redirections in there, and I don't see anything suspect in the backend logs...

Oh, I just tried the same thing on https://spiffdemo.org and saw the same result! Hmmmm...

mogul commented:
This one is also not working correctly, both locally and on https://spiffdemo.org.
![image](https://github.com/GSA-TTS/spiff-arena/assets/47762/b206baf1-bc73-469a-a079-3070c8cb377a)

I asked about them in the upstream #help Discord channel.

mogul committed:
Silence spurious output when pushing with the manifest

mogul committed:
Explicitly set the memory for each app

mogul committed:
Use the provided endpoint for health check

mogul committed:
Remove a spurious line and rename the backend flask key

We want to make it obvious that this variable is only for the backend app

mogul committed:
Add the connector and ensure the backend can connect

We needed to
* add a network policy to enable the backend to hit the internal route
* add the Cloud Foundry-provided CA cert bundle to the Docker
container at startup time
* ensure Python picks up added CA certs

mogul committed:
Rearrange apps to influence startup order

We don't want any app to fail to start up the first time someone
pushes this manifest. So we start with the connector, then the backend,
then the frontend.

Upstream improvement: loop when there are errors like this

asteel-gsa committed:
Allow the use of API extensions

asteel-gsa commented:
Using the extension, I was able to successfully navigate the API keys section, create a key, and do the API demo as present in the API demo.

<div align="center">
  <a href="https://www.youtube.com/watch?v=6fymiAZuRk4">
<img src="https://img.youtube.com/vi/6fymiAZuRk4/0.jpg" alt="Spiffworkflow API Demo"></a>
</div>

![image](https://github.com/GSA-TTS/spiff-arena/assets/130377221/0d674a99-c757-4060-9201-e3a1d7755c7f)

asteel-gsa committed:
Port changed from 8080 -> 80

asteel-gsa committed:
Spacing fixes

asteel-gsa committed:
add trivy (testing)

dependabot[bot] committed:
Bump ruff from 0.5.5 to 0.8.1 in /spiffworkflow-backend

Bumps [ruff](https://github.com/astral-sh/ruff) from 0.5.5 to 0.8.1.
- [Release notes](https://github.com/astral-sh/ruff/releases)
- [Changelog](https://github.com/astral-sh/ruff/blob/main/CHANGELOG.md)
- [Commits](https://github.com/astral-sh/ruff/compare/0.5.5...0.8.1)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
